site/static/api/device-registry-v1.yaml

```diff
@@ -0,0 +1,724 @@
+                  schema:
+                     $ref: '#/components/schemas/Tenant'
+            required: true
+         responses:
+            201:
+               $ref: '#/components/responses/Created'
+            400:
+               $ref: '#/components/responses/MalformedRequest'
+            401:
+               $ref: '#/components/responses/Unauthorized'
+            403:
+               $ref: '#/components/responses/NotAllowed'
+            409:
+               $ref: '#/components/responses/AlreadyExists'
+
+   /tenants/{tenantName}:
+
+      parameters:
+         - $ref: '#/components/parameters/tenantName'
+
+      post:
+         tags:
+            - tenants
+         summary: Create new tenant
+         operationId: createTenantWithId
+         requestBody:
+            description: New device registration
+            content:
+               application/json:
+                  schema:
+                     $ref: '#/components/schemas/Tenant'
+            required: true
+         responses:
+            201:
+               $ref: '#/components/responses/Created'
+            400:
+               $ref: '#/components/responses/MalformedRequest'
+            401:
+               $ref: '#/components/responses/Unauthorized'
+            403:
+               $ref: '#/components/responses/NotAllowed'
+            409:
+               $ref: '#/components/responses/AlreadyExists'
+
+      get:
+         tags:
+            - tenants
+         summary: Get tenant information
+         operationId: getTenant
+         responses:
+            200:
+               description: operation successful
+               content:
+                  application/json:
+                     schema:
+                        $ref: '#/components/schemas/Tenant'
+               headers:
+                  ETag:
+                     description: Version of the resource
+                     schema:
+                        type: string
+            400:
+               $ref: '#/components/responses/MalformedRequest'
+            401:
+               $ref: '#/components/responses/Unauthorized'
+            404:
+               $ref: '#/components/responses/NotFound'
+
+      put:
+         tags:
+            - tenants
+         summary: Update tenant information
+         operationId: updateTenant
+         parameters:
+            - $ref: '#/components/parameters/resourceVersion'
+         requestBody:
+            description: Tenant information
+            content:
+               application/json:
+                  schema:
+                     $ref: '#/components/schemas/Tenant'
+            required: true
+         responses:
+            204:
+               $ref: '#/components/responses/Updated'
+            400:
+               $ref: '#/components/responses/MalformedRequest'
+            401:
+               $ref: '#/components/responses/Unauthorized'
+            403:
+               $ref: '#/components/responses/NotAllowed'
```

Comment:
remind me again: I get a 403 if I am not allowed to update the tenant but I am allowed to read the tenant, and I get a 404 if I am not allowed to read the tenant as well, right?
If so, I find the advantage of having a separate 403 very questionable in comparison to the confusion this seems to create (with me at least) and effort required to implement this distinction ...

Reply:
being responded with 404 when you know the tenant exists but are not allowed to modify it would be very confusing IMHO. How would you understand that you lack the proper permissions with a 404 ? If I was confronted with this (i.e. can read the tenant fine but given 404 when trying to edit) I would loose a lot of time and probably think the service doesn't work properly

Reply:
> remind me again: I get a 403 if I am not allowed to update the tenant but I am allowed to read the tenant, and I get a 404 if I am not allowed to read the tenant as well, right?

Correct.

I think it is important to make a distinct error here. Assuming you have read access to the object. You are going ahead and try to modify it. Your first do a get (OK) and then to an update, which results in "not found".  That seems weird to me. I think it is more important to prevent confusing the end user, considering the developers.

Of course if you (as implementer of this API) can always do the shortcut of returning 404 in any case, if you think that is appropriate to your implementation.

Reply:
The original comment on that would be: https://github.com/eclipse/hono/pull/1130#issuecomment-487567019